File: Lv.1/ex46.js
//프로그래머스 최소직사각형
//풀이르 2번 3번 초기화를 통해 갈아엎었고 도저히 안되어서 문제를 검색해서 풀이 방향성 힌트만 보고 풀어냈다. 답코드를보진않았다.
function solution(sizes) {
    var answer = 0;
    for(let el of sizes){
        if(el[0] < el[1])
            [el[0], el[1]] = [el[1], el[0]];
    }
    let a_max = 0;
    let b_max = 0;
    for(let i=0; i<sizes.length; i++){
        a_max = Math.max(a_max,sizes[i][0]);
        b_max = Math.max(b_max,sizes[i][1]);
    }
    return a_max*b_max;
}

//다른 답 
//마찬가지로 배열의 첫번째값이 두번째값보다 큰값이 오도록 정렬을 하고 그 정렬된 배열에서 두번째값도 제일 큰값을 골라내는 방식으로
//첫번째값이 두번째값보다 크도록 정렬해야한다는 생각을 해내야 한다는게 핵심이라고 생각한다, 어떻게 생각해내야하는건지 매번 답답할뿐.
//map함수와 forEach에서 () 안에 []로 값을 전달해서 사용가능하다는것을 알게되었다.
function solution(sizes) {
    const rotated = sizes.map(([w, h]) => w < h ? [h, w] : [w, h]);

    let maxSize = [0, 0];
    rotated.forEach(([w, h]) => {
        if (w > maxSize[0]) maxSize[0] = w;
        if (h > maxSize[1]) maxSize[1] = h;
    })
    return maxSize[0]*maxSize[1];
}

File: Lv.0/ex46.java
//문자열 묶기
import java.util.HashMap;
import java.util.Map;


//안풀리다가 방법을 찾아보고 HashMap()을 사용한 방법을 활용, JavaScript로 key-value 쌍으로 풀었기 때문에 방법은 알았지만 Java에서 사용할 방법을
//몰라서 찾아보았음.
public class ex46 {
    public int solution(String[] strArr) {
        Map<Integer,Integer> answer = new HashMap<>();
        
        for(String str : strArr){
            //초기값이 없으면 get시에 null을 반환하므로 getOrDefault()로 없으면 0으로 초기화를하고, 그다음에 1을 늘려줘야한다. 
            answer.put(str.length(), answer.getOrDefault(str.length(),0)+1);
        }
        
        int max = 0;
        for(int count : answer.values()){
            if(max < count)
                max = count;
        }
        
        return max;
    }
}

//다른 깔끔한 풀이, 문자의 길이가 최대 30이라고 했으므로 이게 더 정석에 가까운풀이였을것.
public int solution(String[] strArr) {
    int answer = 0;

    int[] lengArr = new int[31];

    for(int i=0; i<strArr.length; i++) {
        //해당 문자열의 길이의 idx값을 +1
        lengArr[strArr[i].length()]++;
    }

    for(int i=0; i<=30; i++) {
        answer = Math.max(answer, lengArr[i]);
    }

    return answer;
}